### ISSUE TITLE
Dockerimage not working

### ISSUE THREAD

**Issue by MikeHofmann** [pypa/bandersnatch#509]
The dockerimage doesn't work, as it seems to not find the 'default.conf':

```
docker run -it --rm pypa/bandersnatch:latest
Running bandersnatch every 3600s
Config file '/conf/bandersnatch.conf' missing, creating default config.
Please review the config file, then run 'bandersnatch' again.
Could not create config file: [Errno 2] No such file or directory: '/usr/local/lib/python3.8/site-packages/bandersnatch/default.conf'
```

but even when i provide a `/conf/bandersnatch.conf` the image still fails, now with:

```
docker run -it --rm -v $PWD/bandersnatch.conf:/conf/bandersnatch.conf:ro pypa/bandersnatch:latest
Running bandersnatch every 3600s
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/site-packages/bandersnatch/main.py", line 183, in <module>
    exit(main())
  File "/usr/local/lib/python3.8/site-packages/bandersnatch/main.py", line 167, in main
    config = bandersnatch.configuration.BandersnatchConfig(
  File "/usr/local/lib/python3.8/site-packages/bandersnatch/configuration.py", line 33, in __call__
    cls._instances[cls] = super().__call__(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/bandersnatch/configuration.py", line 51, in __init__
    with importlib.resources.path("bandersnatch", "default.conf") as config_path:
  File "/usr/local/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/usr/local/lib/python3.8/importlib/resources.py", line 201, in path
    with open_binary(package, resource) as fp:
  File "/usr/local/lib/python3.8/importlib/resources.py", line 91, in open_binary
    return reader.open_resource(resource)
  File "<frozen importlib._bootstrap_external>", line 988, in open_resource
FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/lib/python3.8/site-packages/bandersnatch/default.conf'
```

**Comment by cooperlees** [comment]
You need to specify you new config in the run command. Otherwise it falls back to the default:
https://github.com/pypa/bandersnatch/blob/master/Dockerfile#L21

I would accept a Pull Request to make runner.py smarter in looking for /conf/bandersnatch.conf and falling back to default config

**Comment by MikeHofmann** [comment]
don't think that's the issue. I'll added in the second run via `-v $PWD/bandersnatch.conf:/conf/bandersnatch.conf` a config to the default location where https://github.com/pypa/bandersnatch/blob/master/src/runner.py#L16 is expecting it.

Thus modifying the run command doesn't help either:

`docker run -it --rm -v $PWD/bandersnatch.conf:/conf/bandersnatch.conf pypa/bandersnatch:latest python /bandersnatch/src/runner.py 3600 -c /conf/bandersnatch.conf`

`Running bandersnatch every 3600s
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/site-packages/bandersnatch/main.py", line 183, in <module>
    exit(main())
  File "/usr/local/lib/python3.8/site-packages/bandersnatch/main.py", line 167, in main
    config = bandersnatch.configuration.BandersnatchConfig(
  File "/usr/local/lib/python3.8/site-packages/bandersnatch/configuration.py", line 33, in __call__
    cls._instances[cls] = super().__call__(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/bandersnatch/configuration.py", line 51, in __init__
    with importlib.resources.path("bandersnatch", "default.conf") as config_path:
  File "/usr/local/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/usr/local/lib/python3.8/importlib/resources.py", line 201, in path
    with open_binary(package, resource) as fp:
  File "/usr/local/lib/python3.8/importlib/resources.py", line 91, in open_binary
    return reader.open_resource(resource)
  File "<frozen importlib._bootstrap_external>", line 988, in open_resource
FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/lib/python3.8/site-packages/bandersnatch/default.conf'
`

I believe the problem is, that bandersnatch is looking for default.conf in `/usr/local/lib/python3.8/site-packages/bandersnatch/default.conf` while the file inside the dockerimage is located in `/bandersnatch/src/bandersnatch`

and `docker run -it --rm pypa/bandersnatch:latest find / -name "default.conf"` yields only:

`/bandersnatch/src/bandersnatch/default.conf`

**Comment by cooperlees** [comment]
Ahhh this is been reported before but I was unable to reproduce it locally. It must be the way GitHub actions builds the containers, and that some causes the default.conf to not make it to site packages directory. If you checkout the repo and built the container yourself, do you get the default.conf in the right location? If not, please share how you built it.

I guess we need to try tweak setup.cfg to ensure it always gets included or change that config logic for the docket container.

**Comment by MikeHofmann** [comment]
A colleague wrote a `Dockerfile` which i changed. Main reason to not use your Dockerfile is the `runner.py`. We are somewhat bandwidth-limited and i wanted more control over when mirroring takes place. Especially during these crazy times, i wanted to save bandwidth for video-conferencing and collaboration. So our container uses `cron` for scheduling and some bash-scripts. I can share, but it's to far from your original idea to simple do a pull request.

Reason i noticed the problem with your `Dockerfile`, was that i wanted to simplify the build even further. Atm i build `FROM ubuntu:bionic`, install `python`, `cron`, etc. and then use `pip` to install `bandersnatch`. Also employing a virtualenv for `python`.

```
ENV VIRTUAL_ENV=/opt/venv
RUN python3 -m virtualenv --python=/usr/bin/python3 $VIRTUAL_ENV
ENV PATH="$VIRTUAL_ENV/bin:$PATH"

ADD requirements.txt /opt/venv/
RUN pip install --trusted-host pypi.org --trusted-host pypi.python.org --upgrade pip \
 && pip install --trusted-host pypi.org --trusted-host pypi.python.org --upgrade -r ${VIRTUAL_ENV}/requirements.txt \
 && pip install --trusted-host pypi.org --trusted-host pypi.python.org bandersnatch
```

Basically i wanted to skip the virtualenv and `pip` part and change my sourceimage to `FROM pypa/bandersnatch:latest`, just `apt-get install cron` and be finished (well a few more lines). And the best part would be, that our CICD-Pipeline to build `bandersnatch` as a docker-image, would no longer rely on having a working pipy-mirror (some sort of a Hen-Egg problem)

**Comment by cooperlees** [comment]
I designed the docker image to be used exactly like that (I don't personally use it cause we don't Docker @ Facebook). If we can fix the `default.conf` issue then it should work fine for you. `runner.py` was written to just a simple example of how the container could be used.

You're also welcome to extend the `runner.py`. An idea, that drops the cron dependency is to just make `runner.py` use [python-crontab](https://pypi.org/project/python-crontab/) maybe? and fall back to it's simple mode if there isn't crontab file supplied via a bind mount.

**Comment by SanketDG** [comment]
> It must be the way GitHub actions builds the containers, and that some causes the default.conf to not make it to site packages directory.

This seems strange indeed, the only relevant thing that's probably causing this https://github.com/elgohr/Publish-Docker-Github-Action#cache

Do we do add `no-cache` cron build as suggested (oh this sounds so wrong)?

**Comment by MikeHofmann** [comment]
> You're also welcome to extend the `runner.py`. An idea, that drops the cron dependency is to just make `runner.py` use [python-crontab](https://pypi.org/project/python-crontab/) maybe? and fall back to it's simple mode if there isn't crontab file supplied via a bind mount.

That seems to be a wrapper for systems crontab only. So it wouldn't drop the `cron` dependency. But `cron` is just fine for me. I'll try some things on a build machine tomorrow.

**Comment by MikeHofmann** [comment]
> > It must be the way GitHub actions builds the containers, and that some causes the default.conf to not make it to site packages directory.
> 
> This seems strange indeed, the only relevant thing that's probably causing this https://github.com/elgohr/Publish-Docker-Github-Action#cache
> 
> Do we do add a `no-cache` cron build as suggested?

I just cloned the repo and did a `docker build . --no-cache`. The resulting images produces the same error message.

**Comment by MikeHofmann** [comment]
OK there is a difference in the installed files between the dockerimage vs a install with `pip`.

If i install `bandersnatch` via `pip install --upgrade bandersnatch==4.0.3` 

```
ls -al /usr/local/lib/python3.8/site-packages/bandersnatch
total 128
drwxr-xr-x 3 root root  4096 May 13 07:08 .
drwxr-xr-x 1 root root  4096 May 13 07:08 ..
-rw-r--r-- 1 root root   575 May 13 07:08 __init__.py
drwxr-xr-x 2 root root  4096 May 13 07:08 __pycache__
-rw-r--r-- 1 root root  1929 May 13 07:08 configuration.py
-rw-r--r-- 1 root root  3408 May 13 07:08 default.conf
-rw-r--r-- 1 root root  4317 May 13 07:08 delete.py
-rw-r--r-- 1 root root  5888 May 13 07:08 filter.py
-rw-r--r-- 1 root root   594 May 13 07:08 log.py
-rw-r--r-- 1 root root  5479 May 13 07:08 main.py
-rw-r--r-- 1 root root  6626 May 13 07:08 master.py
-rw-r--r-- 1 root root 20329 May 13 07:08 mirror.py
-rw-r--r-- 1 root root 17424 May 13 07:08 package.py
-rw-r--r-- 1 root root  3645 May 13 07:08 unittest.conf
-rw-r--r-- 1 root root  5055 May 13 07:08 utils.py
-rw-r--r-- 1 root root  8242 May 13 07:08 verify.py
```

vs the dockerimage:

```
docker run -it --rm pypa/bandersnatch:latest ls -al /usr/local/lib/python3.8/site-packages/bandersnatch
total 120
drwxr-xr-x 3 root root  4096 May 11 21:48 .
drwxr-xr-x 1 root root  4096 May 11 21:48 ..
-rw-r--r-- 1 root root   575 May 11 21:47 __init__.py
drwxr-xr-x 2 root root  4096 May 11 21:48 __pycache__
-rw-r--r-- 1 root root  1929 May 11 21:47 configuration.py
-rw-r--r-- 1 root root  4317 May 11 21:47 delete.py
-rw-r--r-- 1 root root  5888 May 11 21:47 filter.py
-rw-r--r-- 1 root root   594 May 11 21:47 log.py
-rw-r--r-- 1 root root  5479 May 11 21:47 main.py
-rw-r--r-- 1 root root  6626 May 11 21:47 master.py
-rw-r--r-- 1 root root 20329 May 11 21:47 mirror.py
-rw-r--r-- 1 root root 17424 May 11 21:47 package.py
-rw-r--r-- 1 root root  5055 May 11 21:47 utils.py
-rw-r--r-- 1 root root  8242 May 11 21:47 verify.py
```

the .conf files are all missing.

**Comment by SanketDG** [comment]
> I just cloned the repo and did a `docker build . --no-cache`. The resulting images produces the same error message.

Did the same, the conf files seem to be there

```
➜ sudo docker run -it --rm 7fcc6d ls -al /usr/local/lib/python3.8/site-packages/bandersnatch
total 124
drwxr-xr-x 3 root root  4096 May 13 11:03 .
drwxr-xr-x 1 root root  4096 May 13 11:03 ..
-rw-r--r-- 1 root root   575 May 13 10:49 __init__.py
drwxr-xr-x 2 root root  4096 May 13 11:03 __pycache__
-rw-r--r-- 1 root root  1929 Apr  5 18:13 configuration.py
-rw-rw-r-- 1 root root  3408 Apr  2 16:02 default.conf
-rw-r--r-- 1 root root  4317 Apr  2 16:02 delete.py
-rw-r--r-- 1 root root  5888 Apr  5 18:13 filter.py
-rw-r--r-- 1 root root   594 Apr  2 16:02 log.py
-rw-r--r-- 1 root root  5479 May 13 10:49 main.py
-rw-r--r-- 1 root root  6626 May 13 10:49 master.py
-rw-r--r-- 1 root root 20329 Apr  2 16:02 mirror.py
-rw-r--r-- 1 root root 17424 Apr  5 18:13 package.py
-rw-rw-r-- 1 root root  3645 Apr  2 16:02 unittest.conf
-rw-r--r-- 1 root root  5055 Apr  5 18:13 utils.py
-rw-r--r-- 1 root root  8242 May 13 10:49 verify.py
```

The upstream image has no `.conf` files, so this definitely seems like an issue on the built upstream image :confused: 

```
sudo docker run -it --rm pypa/bandersnatch ls -al /usr/local/lib/python3.8/site-packages/bandersnatch
total 116
drwxr-xr-x 3 root root  4096 May 12 15:13 .
drwxr-xr-x 1 root root  4096 May 12 15:13 ..
-rw-r--r-- 1 root root   575 May 12 15:12 __init__.py
drwxr-xr-x 2 root root  4096 May 12 15:13 __pycache__
-rw-r--r-- 1 root root  1929 May 12 15:12 configuration.py
-rw-r--r-- 1 root root  4317 May 12 15:12 delete.py
-rw-r--r-- 1 root root  5888 May 12 15:12 filter.py
-rw-r--r-- 1 root root   594 May 12 15:12 log.py
-rw-r--r-- 1 root root  5479 May 12 15:12 main.py
-rw-r--r-- 1 root root  6626 May 12 15:12 master.py
-rw-r--r-- 1 root root 20329 May 12 15:12 mirror.py
-rw-r--r-- 1 root root 17424 May 12 15:12 package.py
```

**Comment by MikeHofmann** [comment]
:confused: as well, as i'm getting different results:

**`git clone https://github.com/pypa/bandersnatch.git`**
```
Cloning into 'bandersnatch'...
remote: Enumerating objects: 60, done.
remote: Counting objects: 100% (60/60), done.
remote: Compressing objects: 100% (54/54), done.
remote: Total 4548 (delta 25), reused 15 (delta 5), pack-reused 4488
Receiving objects: 100% (4548/4548), 826.62 KiB | 2.16 MiB/s, done.
Resolving deltas: 100% (2927/2927), done.
```

**`cd bandersnatch && docker build . --quiet`**
```
sha256:06a75a197c2bf271c9482c1481a725be28a48c06fc812b3c979865e5b77a23f8
```

**`docker run -it --rm `**
```
sha256:06a75a197c2bf271c9482c1481a725be28a48c06fc812b3c979865e5b77a23f8 ls -l /usr/local/lib/python3.8/site-packages/bandersnatch
total 108
-rw-r--r-- 1 root root   575 May 13 06:51 __init__.py
drwxr-xr-x 2 root root  4096 May 13 07:30 __pycache__
-rw-r--r-- 1 root root  1929 May 13 06:51 configuration.py
-rw-r--r-- 1 root root  4317 May 13 06:51 delete.py
-rw-r--r-- 1 root root  5888 May 13 06:51 filter.py
-rw-r--r-- 1 root root   594 May 13 06:51 log.py
-rw-r--r-- 1 root root  5479 May 13 06:51 main.py
-rw-r--r-- 1 root root  6626 May 13 06:51 master.py
-rw-r--r-- 1 root root 20329 May 13 06:51 mirror.py
-rw-r--r-- 1 root root 17424 May 13 06:51 package.py
-rw-r--r-- 1 root root  5055 May 13 06:51 utils.py
-rw-r--r-- 1 root root  8242 May 13 06:51 verify.py
```

**Comment by cooperlees** [comment]
Ok - this annoyed me so I merged #519 

Seems to be good:
- `docker pull pypa/bandersnatch`
- `docker run -d pypa/bandersnatch sleep 600`
- `docker exec -it HASH bash`
```
root@5a5ae4eedbb9:/# find /usr/local/lib/python3.8/ -type f -name '*.conf'
/usr/local/lib/python3.8/site-packages/bandersnatch/unittest.conf
/usr/local/lib/python3.8/site-packages/bandersnatch/default.conf
/usr/local/lib/python3.8/site-packages/oslo_config/tests/test_fixture.conf
```

*(No idea what oslo_config is)*

Can you try again?

**Comment by MikeHofmann** [comment]
Trying:

**`git pull`**
```
remote: Enumerating objects: 59, done.
remote: Counting objects: 100% (59/59), done.
remote: Compressing objects: 100% (37/37), done.
remote: Total 67 (delta 32), reused 26 (delta 22), pack-reused 8
Unpacking objects: 100% (67/67), 120.41 KiB | 948.00 KiB/s, done.
From https://github.com/pypa/bandersnatch
   7c6d1ec..6c7d7f6  master     -> origin/master
...
```
**`docker build --no-cache --quiet .`**
```
sha256:8769cd2deb8ee260b788c992f3c87cf187db6e95105e10b88215692964cb1c87
```
**`docker run -it --rm sha256:8769cd2deb8ee260b788c992f3c87cf187db6e95105e10b88215692964cb1c87 ls -l /usr/local/lib/python3.8/site-packages/bandersnatch`**
```
total 136
-rw-r--r-- 1 root root   579 May 18 11:08 __init__.py
drwxr-xr-x 2 root root  4096 May 18 11:09 __pycache__
-rw-r--r-- 1 root root  1927 May 18 11:08 configuration.py
-rw-rw-r-- 1 root root  3437 May 18 11:08 default.conf
-rw-r--r-- 1 root root  4470 May 18 11:08 delete.py
-rw-r--r-- 1 root root  5888 May 13 12:07 filter.py
-rw-r--r-- 1 root root   594 May 13 12:07 log.py
-rw-r--r-- 1 root root  5677 May 18 11:08 main.py
-rw-r--r-- 1 root root  6626 May 13 12:07 master.py
-rw-r--r-- 1 root root 23327 May 18 11:08 mirror.py
-rw-r--r-- 1 root root 17902 May 18 11:08 package.py
-rw-r--r-- 1 root root 12348 May 18 11:08 storage.py
-rw-rw-r-- 1 root root  3861 May 18 11:08 unittest.conf
-rw-r--r-- 1 root root  4232 May 18 11:08 utils.py
-rw-r--r-- 1 root root  8235 May 18 11:08 verify.py
```

looks good :+1: 

a short test proves that this works.
